clients/drcachesim/tools/filter/trim_filter.h

```diff
@@ -0,0 +1,111 @@
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef _TRIM_FILTER_H_
+#define _TRIM_FILTER_H_ 1
+
+#include "record_filter.h"
+#include "trace_entry.h"
+
+#include <limits>
+#include <vector>
+#include <unordered_set>
+
+namespace dynamorio {
+namespace drmemtrace {
+
+class trim_filter_t : public record_filter_t::record_filter_func_t {
+public:
+    trim_filter_t(uint64_t trim_before_timestamp, uint64_t trim_after_timestamp)
+        : trim_before_timestamp_(trim_before_timestamp)
+        , trim_after_timestamp_(trim_after_timestamp)
+    {
+        // Support 0 to make it easier for users to have no trim-after.
+        if (trim_after_timestamp_ == 0) {
+            trim_after_timestamp_ = (std::numeric_limits<uint64_t>::max)();
+        }
+        if (trim_after_timestamp_ <= trim_before_timestamp_) {
+            error_string_ = "Invalid parameters: end must be > start";
+        }
+    }
+    void *
+    parallel_shard_init(memtrace_stream_t *shard_stream,
+                        bool partial_trace_filter) override
+    {
+        per_shard_t *per_shard = new per_shard_t;
+        return per_shard;
+    }
+    bool
+    parallel_shard_filter(trace_entry_t &entry, void *shard_data) override
+    {
+        per_shard_t *per_shard = reinterpret_cast<per_shard_t *>(shard_data);
+        if (entry.type == TRACE_TYPE_MARKER &&
+            entry.size == TRACE_MARKER_TYPE_TIMESTAMP) {
+            if (entry.addr < trim_before_timestamp_ || entry.addr > trim_after_timestamp_)
```

Comment:
We discussed offline in the past the problems with trimming by instruction count: it is very hard to line up all the threads.  The timestamp cuts across all threads and we decided is the best choice for general trimming.  There may be a use case for trimming a single thread by just a small amount and not worrying about its alignment with other threads?  If we hit that use case we can extend the tool at that point.  It seems like a risky tool though: easy to mess up inter-thread timing.

Reply:
Oh I wasn't suggesting trimming by instr count or making any changes. Just that we note in a comment perhaps that this may remove slightly more entries from the trace due to the understandably limited granularity of timestamps.

Reply:
Added a comment at the top of the trim_filter_t class.